pwiz_tools/Skyline/Executables/BuildMethod/BuildWatersMethod/Program.cs

```diff
@@ -323,6 +325,197 @@ public void build()
 
                 if (!File.Exists(methodTranList.OutputMethod))
                     throw new IOException(string.Format("Failure creating method file {0}.", methodTranList.FinalMethod));
+                UpdatePolarities(methodTranList);
+            }
+        }
+
+        private struct PrecursorInfo
+        {
+            public double? rt;
+            public double? rt_window;
+            public bool is_negative;
+
+            public double? rt_start
+            {
+                get
+                {
+                    if (rt.HasValue && rt_window.HasValue)
+                        return Math.Max(0, rt.Value - rt_window.Value / 2);
+                    return null;
+                }
+            }
+
+            public double? rt_end
+            {
+                get
+                {
+                    if (rt.HasValue && rt_window.HasValue)
+                        return rt + rt_window / 2;
+                    return null;
+                }
+            }
+        }
+
+        private void UpdatePolarities(MethodTransitions methodTransList)
+        {
+            // The algorithm for updating polarities is as follows:
+            // Find all the unique precursors in the transitions list, and for each one
+            // Find a Function section in the method file that has that precursor name and m/z in the CompoundName_1 field
+            var precursorInfo = new Dictionary<string, PrecursorInfo>();
+            using (var transitionsReader = new DsvStreamReader(new StringReader(methodTransList.TransitionList), ','))
+            {
+                while (!transitionsReader.EndOfStream)
+                {
+                    transitionsReader.ReadLine();
+                        var precursorId = transitionsReader["protein.name"] + "," +
+                                          transitionsReader["peptide.seq"] + "," +
+                                          transitionsReader["precursor.mz"];
+                    if (!precursorInfo.ContainsKey(precursorId))

```

Comment:
Do we know how this plays with non-proteomic data?

Reply:
The Skyline documents they provided for testing are actually small molecules. Maybe I should look for some proteomics data to test it on both.

Reply:
Probably a good idea.

Reply:
It looks like Skyline does not allow negative charges in proteomics mode. I guess I should still make sure that RT windows are written correctly.

Reply:
Right, our model of how peptides fragment is specific to protonation so we have that as a basic assumption. No deprotonation of any other weird adducts.